Add Cross-Origin Storage blog post (#3433)
huggingface/blog: cross-origin-storage.md

---
title: "Experimenting with the proposed Cross-Origin Storage API in Transformers.js"
thumbnail: /blog/assets/cross-origin-storage/thumbnail.jpg
authors:
- user: tomayac

>   guest: true

Reviewer: ```suggestion
  guest: true
  org: google
```

This _could_ display a Google logo under the author's name (but [they are not a member](https://huggingface.co/tomayac) of the Google org, so not fully sure if it'll work).

---

# Experimenting with the proposed Cross-Origin Storage API in Transformers.js

(This is a guest post by Developer Relations Engineer [Thomas Steiner](https://blog.tomayac.com/) from the Chrome team at Google.)

Transformers.js provides Web developers with a simple way to use the power of transformers in their Web apps through task-specific pipelines. To run inference in the browser, developers create an instance of [`pipeline()`](https://huggingface.co/docs/transformers.js/en/api/pipelines) and specify a task they want to use the pipeline for. As a concrete example, the following snippet shows how to set up an automatic speech recognition (ASR) pipeline.

```js
import { pipeline } from 'https://cdn.jsdelivr.net/npm/@huggingface/transformers@4.2.0';

const asr = await pipeline(
  'automatic-speech-recognition',
  'Xenova/whisper-tiny.en',
  { device: 'webgpu' },
);
const result = await asr('jfk.wav');
console.log(result);
```

![A minimalistic example of the automatic speech recognition pipeline.](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/cross-origin-storage/87a91qnbicf.png)

## The cache challenge

> You will notice in the source code that I specified [`Xenova/whisper-tiny.en`](https://huggingface.co/Xenova/whisper-tiny.en) as the model, which is a very decent choice for common English automatic speech recognition tasks. In fact, it's even _the_ default model according to the Transformers.js [default model resolution](https://github.com/huggingface/transformers.js/blob/main/packages/transformers/src/pipelines/index.js), as per the following [excerpt](https://github.com/huggingface/transformers.js/blob/bc9cf7400f4f2c8695016699f879e31026ff0313/packages/transformers/src/pipelines/index.js#L151-L158).

Author: ```suggestion
You will notice in the source code that I specified [`Xenova/whisper-tiny.en`](https://huggingface.co/Xenova/whisper-tiny.en) as the model, which is a very decent choice for common English automatic speech recognition tasks. In fact, it's even _the_ default model according to the Transformers.js [default model resolution](https://github.com/huggingface/transformers.js/blob/main/packages/transformers/src/pipelines/index.js), as per the linked [excerpt](https://github.com/huggingface/transformers.js/blob/bc9cf7400f4f2c8695016699f879e31026ff0313/packages/transformers/src/pipelines/index.js#L151-L158).
```

However, `Xenova/whisper-tiny.en` being a popular model (and, as mentioned before, even being _the_ ASR default model in Transformers.js), you can well imagine that more than just one app that you visit would use it. To simulate this situation, here's the same example app from before, but served from a [different origin](https://rawcdn.rawgit.net/GoogleChrome/samples/c4192bd7a3c66fc288a7b22b77acb935df00b8a1/transformersjs-automatic-speech-recognition/index.html). When you visit this different origin app, rather than being usable almost instantly, the browser instead has to download and cache all the model resources again, even if they're byte-by-byte the same as before. Even in this toy example, this adds up to 177 MB of duplicate download and storage, as you can examine in the **Storage** section of the Chrome DevTools [Application panel](https://developer.chrome.com/docs/devtools/application#open_the_application_panel). You can imagine that this quickly adds up.

> ![The Chrome DevTools Storage overview showing 177 MB of used storage.](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/cross-origin-storage/9byoniem0pw.png)

Reviewer: But the original model is ~150 MB if I understand correctly ([encoder model + decoder_merged](https://huggingface.co/Xenova/whisper-tiny.en/tree/main/onnx)). Shouldn't we be seeing ~300 MB occupation in the cache, if it has been cached twice?

Author: The view is per-origin, so you see 150MB on origin A and the same 150MB on origin B. There's no combined view, but if there were, we'd indeed see 300MB.

Two entirely different AI models, but they depend on the same 4,733 kB `ort-wasm-simd-threaded.asyncify.wasm` WebAssembly (Wasm) runtime file [from the underlying ONNX Runtime library](https://onnxruntime.ai/docs/api/js/interfaces/Env.WasmFilePaths.html#wasm) that Transformers.js is built on top of. Open the [extended demo on a different origin](https://rawcdn.rawgit.net/GoogleChrome/samples/d47114a15637383015c274e7bdcd81e1a17b0ccf/transformersjs-automatic-speech-recognition/index2.html), and you will notice in the [**Network** tab](https://developer.chrome.com/docs/devtools/network#load) how also the Wasm runtime gets downloaded and cached again.

![Chrome DevTools Network panel showing the download of the Wasm runtime resource.](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/cross-origin-storage/pz12g20fqeg.png)

So even if you run apps that don't share the same AI models, your browser still makes redundant requests for shared Wasm resources you already have, and on top of that also caches them again, which consumes space on your hard disk.

### Cache isolation

#### AI model resources serving

> By default, **AI model resources** come from the [Hugging Face Hub](https://huggingface.co/docs/hub/en/models-the-hub), and ultimately the Hugging Face CDN. The browser makes a request for a resource like [`https://huggingface.co/Xenova/distilbert-base-uncased-finetuned-sst-2-english/resolve/main/config.json`](https://huggingface.co/Xenova/distilbert-base-uncased-finetuned-sst-2-english/resolve/main/config.json) which then gets redirected to the final CDN URL like [`https://huggingface.co/api/resolve-cache/models/Xenova/distilbert-base-uncased-finetuned-sst-2-english/0b6928efcb76139cae2c6881d49cda67fe119f42/config.json?%2FXenova%2Fdistilbert-base-uncased-finetuned-sst-2-english%2Fresolve%2Fmain%2Fconfig.json=&etag=%223c36342ef1f74de2797d667c68c6b7b988d0b87c%22`](https://huggingface.co/api/resolve-cache/models/Xenova/distilbert-base-uncased-finetuned-sst-2-english/0b6928efcb76139cae2c6881d49cda67fe119f42/config.json?%2FXenova%2Fdistilbert-base-uncased-finetuned-sst-2-english%2Fresolve%2Fmain%2Fconfig.json=&etag=%223c36342ef1f74de2797d667c68c6b7b988d0b87c%22) in this case.

Reviewer: Perhaps no need to enable the links in this case and just show the raw URLs instead (readers don't have to link to understand it, I guess).

Author: I am sure someone will click through just to prove what I wrote is true. I made sure the links are harmless and don't download Gigabytes of data.

## Enter the Cross-Origin Storage API

> **💡 Note:** The Cross-Origin Storage API is an early-stage proposal that isn't final. While the proposed API is not yet natively implemented in any browser, you don't have to wait to experiment with it. Install the [Cross-Origin Storage extension](https://chromewebstore.google.com/detail/cross-origin-storage/denpnpcgjgikjpoglpjefakmdcbmlgih) to inject the `navigator.crossOriginStorage` polyfill on all pages and test the complete flow.

The proposed **[Cross-Origin Storage](https://github.com/WICG/cross-origin-storage) (COS) API** introduces a dedicated `navigator.crossOriginStorage` interface through which web apps can store and retrieve large files across origin boundaries, identified not by a URL, but by a cryptographic hash.

<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/cross-origin-storage/klwb5fryaa.png" alt="The Cross-Origin Storage API logo: a stylized walking person, as typically encountered on crosswalk signs." width="200" height="200">

> That last point about cryptographic hashes is key. Because COS identifies files by their **hash** rather than by their URL or origin, the same `ort-wasm-simd-threaded.asyncify.wasm` Wasm runtime you downloaded while visiting `https://googlechrome.github.io` is recognized as identical to the one `https://rawcdn.rawgit.net` is about to request, no matter where either of the two origins fetched it from. See the following code snippet for the basic flow.

Reviewer: ```suggestion
That last point about cryptographic hashes is key. Because COS identifies files by their **hash** rather than by their URL or origin, the same `ort-wasm-simd-threaded.asyncify.wasm` Wasm runtime you downloaded while visiting `https://googlechrome.github.io` is recognized as identical to the one `https://rawcdn.rawgit.net` is about to request, no matter where either of the two origins fetched it from. See the following code snippet that illustrates the basic flow.
```

This is just meant as a description of the algorithm, right? (not something users / developers would have to build or incorporate)

Crucially, this means an error is not a definitive answer. It might mean "not stored", or it might mean "stored, but the browser isn't telling you". Apps should always handle it the same way: fall back to the network.

### What this means for the Transformers.js example

Going back to the toy examples from before: the `ort-wasm-simd-threaded.asyncify.wasm` runtime weighs in at 4,733 kB and is shared by every Transformers.js-powered app regardless of which AI model it uses. With COS, the first app to load it downloads it once and stores it under its SHA-256 hash with `origins: '*'`. Every subsequent app, whether on `https://googlechrome.github.io`, on `https://rawcdn.rawgit.net`, or any other origin, finds it in COS immediately. The 177 MB of duplicate Whisper model weights? Same story: `Xenova/whisper-tiny.en` gets downloaded once, recognized by hash the second time around, and served from COS in milliseconds. And of course, the same also happens for  `Xenova/distilbert-base-uncased-finetuned-sst-2-english`.

> Transformers.js itself is already piloting the COS API at the library level. [Pull request #1549](https://github.com/huggingface/transformers.js/pull/1549) introduced an experimental COS cache backend behind an opt-in flag. Enabling it takes a single line before you set up your pipeline:

Reviewer: Am I understanding it right in that the PR requires the COS extension to be installed?
Are there any other requirements (server headers, for instance)?

Author: Nope, just the extension. That's all.

Author: And even the extension is optional. If you don't have it, your code will still work as it does now (with the duplicate downloads and storage downsides, which is the whole motivation for installing the extension, and ultimately having the API natively in the browser).

```js
import { env, pipeline } from "https://cdn.jsdelivr.net/npm/@huggingface/transformers@4.2.0";

// 👇 Opt in to the experimental Cross-Origin Storage cache backend.
env.experimental_useCrossOriginStorage = true;

const asr = await pipeline('automatic-speech-recognition', 'Xenova/whisper-tiny.en', { device: 'webgpu' });
const result = await asr('jfk.wav');
console.log(result);
```

> With that flag set, Transformers.js resolves the SHA-256 hash for each [Xet-tracked](https://huggingface.co/docs/hub/en/xet/index) model file (the large ONNX weight files) by fetching the raw Xet pointer ([example raw pointer file](https://huggingface.co/Xenova/whisper-tiny.en/raw/main/onnx/decoder_model.onnx)) and extracting its `oid sha256:` field. It then uses that hash as the key for `navigator.crossOriginStorage`. If the model is already in COS (because another site stored it there first), it's served instantly without a network round-trip. If not, it falls back to a regular download and stores the result in COS for the next caller. With the toy example, the advantage in practice is that `Xenova/whisper-tiny.en` and `Xenova/distilbert-base-uncased-finetuned-sst-2-english` (and of course `ort-wasm-simd-threaded.asyncify.wasm`) only ever need to cross the ether once, regardless of how many different origins ask for them.

Reviewer: Amazing, this is what I was thinking while reading about hashing above. Very cool to use the pre-computed Xet hash! 🔥

Author: Yepp, it was a very welcome discovery when I was snooping around your hub! It's also what makes the [cos-resource-fetcher](https://github.com/GoogleChromeLabs/web-ai-demos/tree/main/cos-resource-fetcher) super ergonomic when used with Hugging Face resources.

Note the `experimental_` prefix on the flag. It's intentional and signals that the underlying browser API has not yet been standardized and may change without a major version bump.

### Try it today

> The COS API is not yet natively implemented in any browser, but you don't have to wait to experiment with it. Install the [Cross-Origin Storage extension](https://chromewebstore.google.com/detail/cross-origin-storage/denpnpcgjgikjpoglpjefakmdcbmlgih) to inject the `navigator.crossOriginStorage` polyfill on all pages and test the complete flow. You can inspect the [source code of the extension](https://github.com/web-ai-community/cross-origin-storage-extension) and follow the [usage instructions](https://github.com/web-ai-community/cross-origin-storage-extension?tab=readme-ov-file#usage) to get started.

Reviewer: Fills a bit repetitive with the note added to L118; perhaps we can rewrite slightly or refer to this section in the note.

Author: I think it's fine to repeat this, as this is a deep-linkable paragraph and having the repetition makes it self-contained. Works?

![Chrome Web Store page for the Cross-Origin Storage extension.](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/cross-origin-storage/0q3rowmy67ta.png)

> With the extension installed, you can try the full end-to-end experience right now: open the first [toy example with COS enabled](https://googlechrome.github.io/samples/transformersjs-automatic-speech-recognition/index3.html), let it load `Xenova/whisper-tiny.en`, then open the [toy example with COS enabled from the second origin](https://rawcdn.rawgit.net/GoogleChrome/samples/1e4f2b8c10adc394352c6ec8327bb503bac7aba1/transformersjs-automatic-speech-recognition/index3.html). Instead of the 177 MB re-download you saw before, the model is served from COS in milliseconds. When you open the extension's popup window, you can see COS in action. If you **View by Resource**, you can see the resource with the SHA-256 hash `950978b1dbcbf250335358c1236053ba19a7f7849b33dc777f4421b72b7626fa` shared across `https://googlechrome.github.io` and `https://rawcdn.rawgit.net`.  It may not be obvious, but as you can verify by comparing the SHA-256 hash on Hugging Face, you're looking at [`https://huggingface.co/Xenova/whisper-tiny.en/blob/main/onnx/decoder_model_merged.onnx`](https://huggingface.co/Xenova/whisper-tiny.en/blob/main/onnx/decoder_model_merged.onnx). For now, the extension is mostly aimed at power users like you. Once implemented in the browser, there will be a friendlier integration in the browser's **Settings** page. The screenshot below shows the extension's popup window with the **View by Resource** tab active, where you can see the shared resource with its hash and the two origins that have it in their COS cache.

Reviewer: Aha, I see `Access-Control-Allow-Origin: *` coming from both URLs and `Cross-Origin-Resource-Policy: cross-origin`. Is this the way it's supposed to be set up or am I imagining things here?

Author: While similarly named, these headers have nothing to do with COS. These headers are for Cross-Origin Resource Sharing (CORS), the security concept. Cross-Origin Storage (COS, the storage mechanism) relies on resources being accessible through CORS so resources can be put in the COS cache, but that's it.

![A resource seen in the Cross-Origin Storage extension, showing it's shared between two origins.](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/cross-origin-storage/usg5dq7dhm.png)

> ## Call to action
> 
> If you're building your own Transformers.js app, the call to action is simple: add `env.experimental_useCrossOriginStorage = true` before your first `pipeline()` call, install the extension, and watch the duplicate downloads disappear from your Network tab. Every site that opts in makes the experience faster and cheaper for every other site's users. Opting in is completely risk-free: if the COS API isn't supported because the user doesn't have the COS extension installed, the code just falls back to the default path (the [Web Cache](https://developer.mozilla.org/en-US/docs/Web/API/Cache) API).
> 
> On the Chrome team, we're [considering implementing the COS API](https://chromestatus.com/feature/5163371507875840) natively in the browser. As an early stage proposal, we welcome feedback on the API, and the shape of the proposal itself. The [Cross-Origin Storage repository](https://github.com/WICG/cross-origin-storage) is the place to file issues, [express support](https://github.com/WICG/cross-origin-storage/labels/expression%20of%20support), or open PRs.

Author: ```suggestion
If you're building your own Transformers.js app, the call to action is simple: add `env.experimental_useCrossOriginStorage = true` before your first `pipeline()` call, install the extension, and watch the duplicate downloads disappear from your Network tab. Every site that opts in makes the experience faster and cheaper for every other site's users. Opting in is completely risk-free: if the COS API isn't supported because the user doesn't have the COS extension installed, the code just falls back to the default path (the [Web Cache](https://developer.mozilla.org/en-US/docs/Web/API/Cache) API).

Transformers.js is not alone in experimenting with COS. [WebLLM](https://webllm.mlc.ai/) (opt-in, see [documentation](https://webllm.mlc.ai/docs/user/advanced_usage.html#using-cross-origin-storage-cache)) and [wllama](https://github.com/ngxson/wllama) (automatic, see [PR](https://github.com/ngxson/wllama/pull/248)) are likewise excited about this proposed API.

On the Chrome team, we're [considering implementing the COS API](https://chromestatus.com/feature/5163371507875840) natively in the browser. As an early stage proposal, we welcome feedback on the API, and the shape of the proposal itself. The [Cross-Origin Storage repository](https://github.com/WICG/cross-origin-storage) is the place to file issues, [express support](https://github.com/WICG/cross-origin-storage/labels/expression%20of%20support), or open PRs.
```

---
title: "Experimenting with the proposed Cross-Origin Storage API in Transformers.js"
thumbnail: /blog/assets/cross-origin-storage/thumbnail.svg
authors:

> - user: tomayac

Reviewer 2: ```suggestion
- user: tomayac
  guest: true
```

---
title: "Experimenting with the proposed Cross-Origin Storage API in Transformers.js"
thumbnail: /blog/assets/cross-origin-storage/thumbnail.jpg
authors:
- user: tomayac
  guest: true
---

# Experimenting with the proposed Cross-Origin Storage API in Transformers.js

(This is a guest post by Developer Relations Engineer [Thomas Steiner](https://blog.tomayac.com/) from the Chrome team at Google.)

Transformers.js provides Web developers with a simple way to use the power of transformers in their Web apps through task-specific pipelines. To run inference in the browser, developers create an instance of [`pipeline()`](https://huggingface.co/docs/transformers.js/en/api/pipelines) and specify a task they want to use the pipeline for. As a concrete example, the following snippet shows how to set up an automatic speech recognition (ASR) pipeline.

```js

> import { pipeline } from 'https://unpkg.com/@huggingface/transformers@4.2.0/dist/transformers.min.js';

Reviewer 3: ```suggestion
import { pipeline } from 'https://cdn.jsdelivr.net/npm/@huggingface/transformers@4.2.0';
```

can be somewhat simplified. In the README we use jsdelivr, but we can use unpkg too (I'm just saying we shouldn't need `dist/transformers.min.js` as it should be inferred; works in jsdelivr, probably unpkg too)

  { device: 'webgpu' },
);
const result = await asr('jfk.wav');
console.log(result);
```

![A minimalistic example of the automatic speech recognition pipeline.](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/cross-origin-storage/87a91qnbicf.png)

## The cache challenge

You will notice in the source code that I specified [`Xenova/whisper-tiny.en`](https://huggingface.co/Xenova/whisper-tiny.en) as the model, which is a very decent choice for common English automatic speech recognition tasks. In fact, it's even _the_ default model according to the Transformers.js [default model resolution](https://github.com/huggingface/transformers.js/blob/main/packages/transformers/src/pipelines/index.js), as per the following [excerpt](https://github.com/huggingface/transformers.js/blob/bc9cf7400f4f2c8695016699f879e31026ff0313/packages/transformers/src/pipelines/index.js#L151-L158).

```js
{
  /* … */
  'automatic-speech-recognition': {
    pipeline: AutomaticSpeechRecognitionPipeline,
    model: [AutoModelForSpeechSeq2Seq, AutoModelForCTC],
    default: {
      model: 'Xenova/whisper-tiny.en',
    },
    type: 'multimodal',
  },
  /* … */
}

> ```

Reviewer 3: ```suggestion
```

I'd say we can remove this -- as we mentioned already it is the default, and a user can visit the link if they wish to see the resolution data structure. imo not important for a regular reader.

<table>
  <thead>
    <tr>
      <th colspan="2">Network Isolation Key</th>
      <th rowspan="2"><strong>Resource URL</strong></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><strong>Top-level site</strong></td>
      <td><strong>Current-frame site</strong></td>
    </tr>
    <tr>
      <td><p><pre>
https://googlechrome.github.io
</pre></p></td>
      <td><p><pre>
https://googlechrome.github.io
</pre></p></td>
      <td><p><pre>
https://cdn.jsdelivr.net/npm/onnxruntime-web@1.26.0-dev.20260416-b7804b056c/dist/ort-wasm-simd-threaded.asyncify.wasm
</pre></p></td>
    </tr>
    <tr>
      <td><p><pre>
https://rawcdn.rawgit.net
</pre></p></td>
      <td><p><pre>
https://rawcdn.rawgit.net
</pre></p></td>
      <td><p><pre>
https://cdn.jsdelivr.net/npm/onnxruntime-web@1.26.0-dev.20260416-b7804b056c/dist/ort-wasm-simd-threaded.asyncify.wasm
</pre></p></td>
    </tr>
  </tbody>
</table>

>  So even if the resource URLs are exactly the same, since the Network Isolation Keys don't match, there's no cache hit, which means duplicate download and duplicate storage. This is the challenge that the Cross-Origin Storage proposal aims to tackle.

Reviewer 3: ```suggestion
So even if the resource URLs are exactly the same, since the Network Isolation Keys don't match, there's no cache hit, which means duplicate download and duplicate storage. This is the challenge that the Cross-Origin Storage proposal aims to tackle.
```

### What this means for the Transformers.js example

Going back to the toy examples from before: the `ort-wasm-simd-threaded.asyncify.wasm` runtime weighs in at 4,733 kB and is shared by every Transformers.js-powered app regardless of which AI model it uses. With COS, the first app to load it downloads it once and stores it under its SHA-256 hash with `origins: '*'`. Every subsequent app, whether on `https://googlechrome.github.io`, on `https://rawcdn.rawgit.net`, or any other origin, finds it in COS immediately. The 177 MB of duplicate Whisper model weights? Same story: `Xenova/whisper-tiny.en` gets downloaded once, recognized by hash the second time around, and served from COS in milliseconds. And of course, the same also happens for  `Xenova/distilbert-base-uncased-finetuned-sst-2-english`.

Transformers.js itself is already piloting the COS API at the library level. [Pull request #1549](https://github.com/huggingface/transformers.js/pull/1549) introduced an experimental COS cache backend behind an opt-in flag. Enabling it takes a single line before you set up your pipeline:

```js

> import { env, pipeline } from "https://unpkg.com/@huggingface/transformers@4.2.0/dist/transformers.min.js";

Author: ```suggestion
import { env, pipeline } from "https://cdn.jsdelivr.net/npm/@huggingface/transformers@4.2.0";
```

> import { env, pipeline } from "https://unpkg.com/@huggingface/transformers@4.2.0/dist/transformers.min.js";

Reviewer 3: same comment about import as above

![A resource seen in the Cross-Origin Storage extension, showing it's shared between two origins.](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/cross-origin-storage/usg5dq7dhm.png)

## Call to action

> If you're building your own Transformers.js app, the call to action is simple: add `env.experimental_useCrossOriginStorage = true` before your first `pipeline()` call, install the extension, and watch the duplicate downloads disappear from your Network tab. Every site that opts in makes the experience faster and cheaper for every other site's users. Opting in is completely risk-free: if the COS API isn't supported because the user doesn't have the COS extension installed, the code just falls back to using the network.

Reviewer 3: ```suggestion
If you're building your own Transformers.js app, the call to action is simple: add `env.experimental_useCrossOriginStorage = true` before your first `pipeline()` call, install the extension, and watch the duplicate downloads disappear from your Network tab. Every site that opts in makes the experience faster and cheaper for every other site's users. Opting in is completely risk-free: if the COS API isn't supported because the user doesn't have the COS extension installed, the code just falls back to the default path (the [Web Cache](https://developer.mozilla.org/en-US/docs/Web/API/Cache) API).
```

Reviewer 3: since i guess we are comparing caching mechanisms; might as well mention what the path is here (not always needing a network request)
